piwik_sdk/src/main/java/org/piwik/sdk/Dispatcher.java

```diff
@@ -0,0 +1,252 @@
+                    e.printStackTrace();
+                }
+
+                int count = 0;
+                List<String> availableEvents = new ArrayList<>();
+                mDispatchQueue.drainTo(availableEvents);
+                Logy.d(LOGGER_TAG, "Drained " + availableEvents.size() + " events.");
+                TrackerBulkURLWrapper wrapper = new TrackerBulkURLWrapper(mApiUrl, availableEvents, mAuthToken);
+                Iterator<TrackerBulkURLWrapper.Page> pageIterator = wrapper.iterator();
+                while (pageIterator.hasNext()) {
+                    TrackerBulkURLWrapper.Page page = pageIterator.next();
+
+                    // use doGET when only event on current page
+                    if (page.elementsCount() > 1) {
+                        if (doPost(wrapper.getApiUrl(), wrapper.getEvents(page)))
+                            count += page.elementsCount();
+                    } else {
+                        if (doGet(wrapper.getEventUrl(page)))
+                            count += 1;
+                    }
+                }
+                Logy.d(LOGGER_TAG, "Dispatched " + count + " events.");
+                synchronized (mThreadControl) {
+                    // We may be done or this was a forced dispatch
+                    if (mDispatchQueue.isEmpty() || mDispatchInterval < 0) {
+                        mRunning = false;
+                        break;
+                    }
+                }
+            }
+        }
+    };
+
+    private boolean doGet(String trackingEndPointUrl) {
+        if (trackingEndPointUrl == null)
+            return false;
+        HttpGet get = new HttpGet(trackingEndPointUrl);
+        return doRequest(get);
+    }
+
+    private boolean doPost(URL url, JSONObject json) {
+        if (url == null || json == null)
+            return false;
+
+        String jsonBody = json.toString();
+        try {
+            HttpPost post = new HttpPost(url.toURI());
+            StringEntity se = new StringEntity(jsonBody);
+            se.setContentType(new BasicHeader(HTTP.CONTENT_TYPE, "application/json"));
+            post.setEntity(se);
+
+            return doRequest(post);
+        } catch (URISyntaxException e) {
+            Logy.w(LOGGER_TAG, String.format("URI Syntax Error %s", url.toString()), e);
+        } catch (UnsupportedEncodingException e) {
+            Logy.w(LOGGER_TAG, String.format("Unsupported Encoding %s", jsonBody), e);
+        }
+        return false;
+    }
+
+    private boolean doRequest(HttpRequestBase requestBase) {
+        HttpClient client = new DefaultHttpClient();
+        HttpConnectionParams.setConnectionTimeout(client.getParams(), mTimeOut);
+        HttpResponse response;
+
+        if (mPiwik.isDryRun()) {
+            Logy.d(LOGGER_TAG, "DryRun, stored HttpRequest, now " + mDryRunOutput.size());
+            mDryRunOutput.add(requestBase);
+        } else {
+            mDryRunOutput.clear();
```

Comment:
I'll rather check if `mDryRUnOutput` is not empty before trying clearing it (mainly bc of [this](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/ArrayList.java#473))